rfc/0078-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,346 @@
+
+```
+class TransactionGetReplicaOptions {
+    Transcoder transcoder;
+}
+```
+
+The method might throw the following errors:
+
+* `101 DocumentNotFound`, when the server returns KV Code `0x01 ENOENT` for
+  *all* requests.
+
+* `102 DocumentIrretrievable`, when the SDK finds that there are nodes in local
+  group, or there is no group available with the name selected in connection
+  options.
+
+# Caveats
+
+The use must be aware of the following topics when dealing with
+Zone-Aware-Replica-Reads:
+
+- Regardless the fact that the Couchbase supports Server Groups for a long time
+  already, the SDK can see them in configuration since 7.6.2 release only, which
+  make Zone-Aware-Replica-Reads to be available only in most recent releases.
+
+- It is crucial to perform rebalance after setting up server groups, as only
+  after this process the data will be physically moved.
+
+- The User must consult server documentation regarding setting up balanced
+  cluster for efficient usage of the feature. Number of the replicas and number
+  of the groups have to be carefully selected.
+
+# Open Questions
+
+## Q1
+
+> What the SDK should do if the local group does not have any replica configured
+> for the key, but the User request `Selected-Server-Group` strategy? Consider
+> the following cases:
+>
+> - number of replicas is 1, number of zones is 3. Neither active, nor replica
+>   is in the local group.
+> - number of replicas is 1, number of zones is 3. Only active node is in the
+>   local group.
+> - number of replicas is 1, number of zones is 3. Only replica node is in the
+>   local group.
+>
+> Shall we warn users when locality restriction turns request into just GET,
+> because only active node is in local group?
+
+The SDK should just implement two strategy, described in the body of the RFC.
+Although future editions might add more strategies with retrying with full
+replica-set or performing extra validations.
+
+## Q2
+
+> Do we need to implement management API to create/retrieve server groups and
+> change node belonging to particular group?
+
+No, we do not need to implement management API. If necessary, REST API should be
+used
+(https://docs.couchbase.com/server/current/manage/manage-groups/manage-groups.html)
+`cbdinocluster` also support server groups since version 0.0.43.
+
+## Q3
+
+> All currently supported server versions allow to configure server groups, but
+> only 7.6.1 announce this information in configuration.
+> [MB-60835](https://issues.couchbase.com/browse/MB-60835) How the SDK should
+> behave with older server? Do we need to fail fast or silently fall back to
+> current behavior?
+
+The behavior of this feature with older servers (pre 7.6.2) is not defined. It
+is up to SDK to decide what to do (for example, it could return `102
+DocumentIrretrievable` in this case, because the configuration does not tell us
+anything about server groups assigned to the nodes)
```

Comment:
I feel we should nail this down.

What you've proposed makes sense to me, in the context of what I propose elsewhere about the user always have try-catch handling.

Reply:
I will update this with reference to the fallback example